File: src/db/pool.py
"""
Async SQLite Connection Pool using aiosqlite

Features:
- Connection pool with configurable size (5-10 connections)
- Context manager for automatic release
- WAL mode for concurrent reads
- Thread-safe singleton pattern
"""

from __future__ import annotations

import asyncio
import logging
from contextlib import asynccontextmanager
from pathlib import Path
from typing import Optional, AsyncIterator, Dict, Any

try:
    import aiosqlite
except ImportError:
    aiosqlite = None

logger = logging.getLogger("niko-db")


class AsyncConnectionPool:
    """
    Async SQLite connection pool with WAL mode support.

    Usage:
        pool = AsyncConnectionPool(db_path, pool_size=5)
        await pool.initialize()

        async with pool.acquire() as conn:
            cursor = await conn.execute("SELECT * FROM memories")
            rows = await cursor.fetchall()

        await pool.close()
    """

    def __init__(
        self,
        db_path: str,
        pool_size: int = 5,
        timeout: float = 30.0
    ):
        self.db_path = Path(db_path)
        self.pool_size = min(max(pool_size, 1), 10)  # Clamp to 1-10
        self.timeout = timeout

        self._pool: asyncio.Queue = asyncio.Queue(maxsize=self.pool_size)
        self._connections: list = []
        self._initialized = False
        self._lock = asyncio.Lock()

    async def initialize(self) -> None:
        """Initialize the connection pool with WAL mode enabled."""
        if aiosqlite is None:
            raise ImportError(
                "aiosqlite is required for AsyncConnectionPool. "
                "Install with: pip install aiosqlite>=0.19.0"
            )

        async with self._lock:
            if self._initialized:
                return

            # Ensure db directory exists
            self.db_path.parent.mkdir(parents=True, exist_ok=True)

            # Create connections
            for i in range(self.pool_size):
                conn = await aiosqlite.connect(
                    str(self.db_path),
                    timeout=self.timeout
                )

                # Enable WAL mode for better concurrent read performance
                await conn.executescript("""
                    PRAGMA journal_mode=WAL;
                    PRAGMA synchronous=NORMAL;
                    PRAGMA cache_size=-64000;
                """)

                # Enable row factory for dict-like access
                conn.row_factory = aiosqlite.Row

                self._connections.append(conn)
                await self._pool.put(conn)

            self._initialized = True
            logger.info(
                f"Connection pool initialized: {self.pool_size} connections "
                f"to {self.db_path}"
            )

    @asynccontextmanager
    async def acquire(self) -> AsyncIterator[aiosqlite.Connection]:
        """
        Acquire a connection from the pool.

        Usage:
            async with pool.acquire() as conn:
                await conn.execute(...)
        """
        if not self._initialized:
            await self.initialize()

        conn = await asyncio.wait_for(
            self._pool.get(),
            timeout=self.timeout
        )

        try:
            yield conn
        finally:
            # Return connection to pool
            await self._pool.put(conn)

    async def execute(self, sql: str, params: tuple = ()) -> aiosqlite.Cursor:
        """Execute a single SQL statement using a pooled connection."""
        async with self.acquire() as conn:
            cursor = await conn.execute(sql, params)
            await conn.commit()
            return cursor

    async def executemany(
        self,
        sql: str,
        params_seq: list
    ) -> aiosqlite.Cursor:
        """Execute SQL with multiple parameter sets."""
        async with self.acquire() as conn:
            cursor = await conn.executemany(sql, params_seq)
            await conn.commit()
            return cursor

    async def executescript(self, script: str) -> None:
        """Execute a SQL script."""
        async with self.acquire() as conn:
            await conn.executescript(script)
            await conn.commit()

    async def fetchall(
        self,
        sql: str,
        params: tuple = ()
    ) -> list:
        """Execute query and fetch all results."""
        async with self.acquire() as conn:
            cursor = await conn.execute(sql, params)
            return await cursor.fetchall()

    async def fetchone(
        self,
        sql: str,
        params: tuple = ()
    ) -> Optional[aiosqlite.Row]:
        """Execute query and fetch one result."""
        async with self.acquire() as conn:
            cursor = await conn.execute(sql, params)
            return await cursor.fetchone()

    async def close(self) -> None:
        """Close all connections in the pool."""
        async with self._lock:
            if not self._initialized:
                return

            # Drain the pool
            while not self._pool.empty():
                try:
                    self._pool.get_nowait()
                except asyncio.QueueEmpty:
                    break

            # Close all connections
            for conn in self._connections:
                try:
                    await conn.close()
                except Exception as e:
                    logger.warning(f"Error closing connection: {e}")

            self._connections.clear()
            self._initialized = False
            logger.info("Connection pool closed")

    @property
    def is_initialized(self) -> bool:
        """Check if pool is initialized."""
        return self._initialized

    @property
    def available_connections(self) -> int:
        """Number of available connections in the pool."""
        return self._pool.qsize()


# Global pool instance (singleton pattern)
_global_pools: Dict[str, AsyncConnectionPool] = {}
_pools_lock = asyncio.Lock()


async def get_pool(
    db_path: str,
    pool_size: int = 5,
    timeout: float = 30.0
) -> AsyncConnectionPool:
    """
    Get or create a connection pool for the given database path.

    This function maintains a singleton pool per database path.

    Args:
        db_path: Path to the SQLite database
        pool_size: Number of connections (1-10)
        timeout: Connection timeout in seconds

    Returns:
        AsyncConnectionPool instance
    """
    db_path = str(Path(db_path).resolve())

    async with _pools_lock:
        if db_path not in _global_pools:
            pool = AsyncConnectionPool(
                db_path,
                pool_size=pool_size,
                timeout=timeout
            )
            await pool.initialize()
            _global_pools[db_path] = pool

        return _global_pools[db_path]


async def close_all_pools() -> None:
    """Close all global connection pools."""
    async with _pools_lock:
        for pool in _global_pools.values():
            await pool.close()
        _global_pools.clear()

File: src-ts/db/pool.ts
/**
 * Database module - Async SQLite Connection Pool
 *
 * Uses better-sqlite3 (synchronous) with a worker-like pool pattern.
 * When pool is exhausted, requests are queued instead of creating
 * temporary connections that bypass pool management.
 */

import BetterSqlite3 from 'better-sqlite3';
import { existsSync, mkdirSync } from 'fs';
import { resolve, dirname } from 'path';

type DatabaseType = InstanceType<typeof BetterSqlite3>;

/** 队列超时时间（毫秒） */
const QUEUE_TIMEOUT_MS = 30_000;

/**
 * 连接池信号量 — 控制并发获取，池耗尽时排队等待
 */
class PoolSemaphore {
  private _queue: Array<{
    resolve: (conn: DatabaseType) => void;
    timer: ReturnType<typeof setTimeout>;
  }> = [];
  private _available: number;

  constructor(max: number) {
    this._available = max;
  }

  /**
   * 获取一个连接。池有空闲时立即返回，否则排队等待。
   * 等待超过 QUEUE_TIMEOUT_MS 后拒绝 Promise。
   */
  acquire(pool: AsyncConnectionPool): Promise<DatabaseType> {
    return new Promise((resolve, reject) => {
      // 池中有空闲连接，直接获取
      if (this._available > 0 && pool['_pool'].length > 0) {
        this._available--;
        resolve(pool['_pool'].pop()!);
        return;
      }

      // 池耗尽，加入等待队列
      const timer = setTimeout(() => {
        const idx = this._queue.findIndex((item) => item.resolve === resolve);
        if (idx !== -1) this._queue.splice(idx, 1);
        reject(new Error('Pool acquire timeout: waited 30s for a connection'));
      }, QUEUE_TIMEOUT_MS);

      this._queue.push({ resolve, timer });
    });
  }

  /**
   * 释放连接。优先分给队列中等待的请求，否则归还可用计数。
   */
  release(conn: DatabaseType, pool: AsyncConnectionPool): void {
    if (this._queue.length > 0) {
      // 有等待中的请求，直接转交连接
      const waiter = this._queue.shift()!;
      clearTimeout(waiter.timer);
      waiter.resolve(conn);
      return;
    }

    // 无等待者，归还到池
    this._available++;
    if (pool['_pool'].length < pool['_poolSize']) {
      pool['_pool'].push(conn);
    } else {
      try { conn.close(); } catch { /* 忽略关闭错误 */ }
    }
  }

  /** 当前等待队列长度 */
  get pendingCount(): number {
    return this._queue.length;
  }

  /** 清空等待队列，拒绝所有等待者 */
  drain(): void {
    for (const item of this._queue) {
      clearTimeout(item.timer);
      item.resolve = () => {}; // 防止悬挂 resolve
    }
    this._queue.length = 0;
  }
}

/**
 * Connection pool for better-sqlite3 databases.
 */
export class AsyncConnectionPool {
  private _dbPath: string;
  private _poolSize: number;
  private _pool: DatabaseType[] = [];
  private _initialized = false;
  private _semaphore: PoolSemaphore;

  constructor(dbPath: string, poolSize: number = 5) {
    this._poolSize = Math.min(Math.max(poolSize, 1), 10);
    this._semaphore = new PoolSemaphore(this._poolSize);
    this._dbPath = dbPath;
  }

  initialize(): void {
    if (this._initialized) return;

    const dir = dirname(this._dbPath);
    if (dir && !existsSync(dir)) {
      mkdirSync(dir, { recursive: true });
    }

    for (let i = 0; i < this._poolSize; i++) {
      const db = new BetterSqlite3(this._dbPath);
      db.pragma('journal_mode = WAL');
      db.pragma('synchronous = NORMAL');
      db.pragma('cache_size = -64000');
      this._pool.push(db);
    }

    this._initialized = true;
  }

  /**
   * 获取一个连接。池耗尽时排队等待而非创建临时连接。
   * 调用方必须在使用完毕后调用 release() 归还连接。
   */
  async acquire(): Promise<DatabaseType> {
    if (!this._initialized) this.initialize();
    return this._semaphore.acquire(this);
  }

  /**
   * 释放连接回池。优先转交给等待队列中的请求。
   */
  release(db: DatabaseType): void {
    this._semaphore.release(db, this);
  }

  close(): void {
    this._semaphore.drain();
    for (const db of this._pool) {
      try { db.close(); } catch { /* 忽略关闭错误 */ }
    }
    this._pool = [];
    this._initialized = false;
    // 重建信号量以重置状态
    this._semaphore = new PoolSemaphore(this._poolSize);
  }

  get isInitialized(): boolean { return this._initialized; }
  get availableConnections(): number { return this._pool.length; }
  get pendingCount(): number { return this._semaphore.pendingCount; }
}

const _globalPools = new Map<string, AsyncConnectionPool>();

export function getPool(dbPath: string, poolSize: number = 5): AsyncConnectionPool {
  const resolvedPath = resolve(dbPath);
  if (!_globalPools.has(resolvedPath)) {
    const pool = new AsyncConnectionPool(resolvedPath, poolSize);
    pool.initialize();
    _globalPools.set(resolvedPath, pool);
  }
  return _globalPools.get(resolvedPath)!;
}

export function closeAllPools(): void {
  for (const pool of _globalPools.values()) pool.close();
  _globalPools.clear();
}
